src/pudl/helpers.py

```diff
@@ -213,187 +213,107 @@ def is_doi(doi):
-        assert delta_year / pd.Timedelta(days=1) <= 366.0
-    elif len(year_index) == 1:
-        assert year_index.min().month == 1, "only month not Jan"
-        assert year_index.min().day == 1, "only day not 1st"
-    else:
-        assert False, "Zero dates found!"
-
-    return True
-
-
-def merge_on_date_year_new(
-    df_date,
-    df_year,
-    date_col="report_date",
-    year_col="report_date",
-    on=None,
-    how="inner",
+def clean_merge_asof(
+    left,
+    right,
+    left_on="report_date",
+    right_on="report_date",
+    by={},
 ):
     """
-    Transition to using pd.merge_asof() instead of our janky homebrew.
-
-    Rather than swapping out all the calls to merge_on_date_year() just yet,
-    this function translates the call signature of that function and uses
-    pd.merge_asof() internally to hopefully simplify and make the process much
-    more robust.
-
-    """
-    left = df_date.copy()
-    left.loc[:, date_col] = pd.to_datetime(left.date_col)
-    right = df_year.copy()
-    right.loc[:, year_col] = pd.to_datetime(right.year_col)
-
-    return (
-        pd.merge_asof(
-            left=left,
-            right=right,
-            left_on=date_col,
-            right_on=year_col,
-            left_by=on,
-            right_by=on,
-        )
-    )
-
-
-def merge_on_date_year(
-    df_date,
-    df_year,
-    on=(),
-    how='inner',
-    date_col='report_date',
-    year_col='report_date',
-):
-    """Merge two dataframes based on a shared year.
-
-    Some of our data is annual, and has an integer year column (e.g. FERC 1).
-    Some of our data is annual, and uses a Date column (e.g. EIA 860), and
-    some of our data has other temporal resolutions, and uses date columns
-    (e.g. EIA 923 fuel receipts are monthly, EPA CEMS data is hourly). This
-    function takes two data frames and merges them based on the year that the
-    data pertains to.  It requires one of the dataframes to have annual
-    resolution, and allows the annual time to be described as either an integer
-    year or a Date. The non-annual dataframe must have a Date column.
-
-    By default, it is assumed that both the date and annual columns to be
-    merged on are called 'report_date' since that's the common case when
-    bringing together EIA860 and EIA923 data.
+    Merge two dataframes having different time report_date frequencies.
+
+    We often need to bring together data which is reported on a monthly basis,
+    and entity attributes that are reported on an annual basis. For example, we
+    might want monthly net generation, and also a bunch of generator attributes
+    in the same dataframe. In addition, we allow aggregation at either monthly
+    or annual resolution in some contexts, and so we need to be able to merge
+    together a monthly or annual data table with an annual entity attribute
+    table. This is exactly what :func:`pandas.merge_asof` is designed to do.
```

Comment:
It seems like the content of this paragraph is generally covered by the first sentence. I 
```suggestion
We often need to bring together data which is reported on a monthly basis,
and entity attributes that are reported on an annual basis. We want to be
flexible in merging monthly and annual; monthly and monthly data; or annual
and annual data. This is exactly what :func:`pandas.merge_asof` is designed
to do.
```

Reply:
Okay I tried to shorten / deduplicate in the docstring generally here.